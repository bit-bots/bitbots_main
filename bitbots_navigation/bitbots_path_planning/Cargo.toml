[package]
name = "bbpprs"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
[lib]
name = "bbpprs"
crate-type = ["cdylib"]

[dependencies]
pyo3 = "0.23.1"
anyhow = "1.0.93"
geo = "0.29.2"
itertools = "0.13.0"
petgraph = "0.6.5"
plotters = "0.3.7"
rand = "0.8.5"
keyed_priority_queue = "0.4.2"
ordered-float = "4.5.0"
