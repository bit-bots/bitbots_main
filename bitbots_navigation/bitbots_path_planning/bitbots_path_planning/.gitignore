path_planning_parameters.py
