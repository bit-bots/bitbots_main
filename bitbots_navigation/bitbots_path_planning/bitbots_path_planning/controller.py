import math

import tf2_ros as tf2
from geometry_msgs.msg import Twist
from nav_msgs.msg import Path
from rclpy.node import Node
from ros2_numpy import numpify
from tf2_geometry_msgs import PointStamped, Pose, PoseStamped
from tf_transformations import euler_from_quaternion


class Controller:
    """
    A simple follow the carrot controller which controls the robots command velocity to stay on a given path.
    """

    def __init__(self, node: Node, buffer: tf2.Buffer) -> None:
        self.node = node
        self.buffer = buffer

        self.config_base_footprint_frame: str = self.node.get_parameter("base_footprint_frame").value
        self.config_carrot_distance: int = self.node.declare_parameter("controller.carrot_distance", 20).value
        self.config_max_rotation_vel: float = self.node.declare_parameter("controller.max_rotation_vel", 0.4).value
        self.config_max_vel_x: float = self.node.declare_parameter("controller.max_vel_x", 0.25).value
        self.config_max_vel_y: float = self.node.declare_parameter("controller.max_vel_y", 0.08).value
        self.config_min_vel_x: float = self.node.declare_parameter("controller.min_vel_x", -0.2).value
        self.config_orient_to_goal_distance: float = self.node.declare_parameter(
            "controller.orient_to_goal_distance", 1.0
        ).value
        self.config_rotation_slow_down_factor: float = self.node.declare_parameter(
            "controller.rotation_slow_down_factor", 0.3
        ).value
        self.config_rotation_i_factor: float = self.node.declare_parameter("controller.rotation_i_factor", 0.05).value
        self.config_smoothing_k: float = self.node.declare_parameter("controller.smoothing_k", 0.2).value
        self.config_translation_slow_down_factor: float = self.node.declare_parameter(
            "controller.translation_slow_down_factor", 0.5
        ).value

        # Last command velocity
        self.last_cmd_vel = Twist()

        # Accumulator for the angular error
        self.angular_error_accumulator = 0

    def step(self, path: Path) -> tuple[Twist, PointStamped]:
        """
        Calculates a command velocity based on a given path
        """
        # Starting point for our goal velocity calculation
        cmd_vel = Twist()

        # Create an default pose in the origin of our base footprint
        pose_geometry_msgs = PoseStamped()
        pose_geometry_msgs.header.frame_id = self.config_base_footprint_frame

        # We get our pose in respect to the map frame (also frame of the path message)
        # by transforming the pose above into this frame
        try:
            current_pose: Pose = self.buffer.transform(pose_geometry_msgs, path.header.frame_id).pose
        except (tf2.ConnectivityException, tf2.LookupException, tf2.ExtrapolationException) as e:
            self.node.get_logger().warn("Failed to perform controller step: " + str(e))
            return cmd_vel

        # Select pose for the carrot (goal_pose for this controller) and the end pose of the global plan
        # End conditions with an empty or shorter than carrot distance path need to be considered
        if len(path.poses) > 0:
            end_pose: Pose = path.poses[-1].pose
            if len(path.poses) > self.config_carrot_distance:
                goal_pose: Pose = path.poses[self.config_carrot_distance].pose
            else:
                goal_pose: Pose = end_pose
        else:
            return cmd_vel

        # Calculate the intended walk angle based on the carrot and our position
        walk_angle = math.atan2(
            goal_pose.position.y - current_pose.position.y, goal_pose.position.x - current_pose.position.x
        )

        # Calculate the heading angle from our current position to the final position of the global plan
        final_walk_angle = math.atan2(
            end_pose.position.y - current_pose.position.y, end_pose.position.x - current_pose.position.x
        )

        # Calculate the distance from our current position to the final position of the global plan
        distance = math.hypot(
            end_pose.position.x - current_pose.position.x, end_pose.position.y - current_pose.position.y
        )

        # Calculate the translational walk velocity.
        # It considers the distance and breaks if we are close to the final position of the global plan
        walk_vel = distance * self.config_translation_slow_down_factor

        # Check if we are so close to the final position of the global plan that we want to align us with
        # its orientation and not the heading towards its position
        diff = 0
        if distance > self.config_orient_to_goal_distance:
            # Calculate the difference between our current heading and the heading towards the final position of
            # the global plan
            diff = final_walk_angle - self._get_yaw(current_pose)
        else:
            # Calculate the difference between our current heading and the heading of the final position of
            # the global plan
            diff = self._get_yaw(end_pose) - self._get_yaw(current_pose)

        # Get the min angle of the difference
        min_angle = math.remainder(diff, math.tau)

        # Accumulate the angular error (integral)
        self.angular_error_accumulator += min_angle

        # Define the target rotation velocity (based on p and i of the error)
        target_rot_vel = (
            self.config_rotation_slow_down_factor * min_angle
            + self.config_rotation_i_factor * self.angular_error_accumulator
        )

        # Clamp rotational velocity to max values (both sides)
        rot_goal_vel = min(max(target_rot_vel, -self.config_max_rotation_vel), self.config_max_rotation_vel)

        # Calculate the x and y components of our linear velocity based on the desired heading and the
        # desired translational velocity.
        cmd_vel.linear.x = math.cos(walk_angle - self._get_yaw(current_pose)) * walk_vel
        cmd_vel.linear.y = math.sin(walk_angle - self._get_yaw(current_pose)) * walk_vel

        # Scale command accordingly if a limit is exceeded
        if cmd_vel.linear.x > self.config_max_vel_x:
            self.node.get_logger().debug(
                f"X max LIMIT reached: {cmd_vel.linear.x} > {self.config_max_vel_x}, with Y being {cmd_vel.linear.y}"
            )
            cmd_vel.linear.y *= self.config_max_vel_x / cmd_vel.linear.x
            cmd_vel.linear.x = self.config_max_vel_x
            self.node.get_logger().debug(f"X max LIMIT reached: scale Y to {cmd_vel.linear.y}")

        if cmd_vel.linear.x < self.config_min_vel_x:
            self.node.get_logger().debug(
                f"X min LIMIT reached: {cmd_vel.linear.x} < {self.config_min_vel_x}, with Y being {cmd_vel.linear.y}"
            )
            cmd_vel.linear.y *= self.config_min_vel_x / cmd_vel.linear.x
            cmd_vel.linear.x = self.config_min_vel_x
            self.node.get_logger().debug(f"X min LIMIT reached: scale Y to {cmd_vel.linear.y}")

        if abs(cmd_vel.linear.y) > self.config_max_vel_y:
            self.node.get_logger().debug(
                f"Y max LIMIT reached: {cmd_vel.linear.y} > {self.config_max_vel_y}, with X being {cmd_vel.linear.x}"
            )
            cmd_vel.linear.x *= self.config_max_vel_y / abs(cmd_vel.linear.y)
            cmd_vel.linear.y = math.copysign(self.config_max_vel_y, cmd_vel.linear.y)
            self.node.get_logger().debug(f"Y max LIMIT reached: scale X to {cmd_vel.linear.x}")

        # Apply the desired rotational velocity
        cmd_vel.angular.z = rot_goal_vel

        # Filter the command velocity to avoid sudden changes
        cmd_vel.linear.x = (
            self.config_smoothing_k * cmd_vel.linear.x + (1 - self.config_smoothing_k) * self.last_cmd_vel.linear.x
        )
        cmd_vel.linear.y = (
            self.config_smoothing_k * cmd_vel.linear.y + (1 - self.config_smoothing_k) * self.last_cmd_vel.linear.y
        )
        cmd_vel.angular.z = (
            self.config_smoothing_k * cmd_vel.angular.z + (1 - self.config_smoothing_k) * self.last_cmd_vel.angular.z
        )

        # Store the last command velocity
        self.last_cmd_vel = cmd_vel

        # Create a carrot message for visualization
        carrot_point = PointStamped()
        carrot_point.header.frame_id = path.header.frame_id
        carrot_point.point = goal_pose.position

        return cmd_vel, carrot_point

    def _get_yaw(self, pose: Pose) -> float:
        """
        Returns the yaw angle of a given pose
        """
        return euler_from_quaternion(numpify(pose.orientation))[2]
