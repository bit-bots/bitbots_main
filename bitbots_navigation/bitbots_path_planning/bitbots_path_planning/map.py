import cv2
import numpy as np
import soccer_vision_3d_msgs.msg as sv3dm
import tf2_ros as tf2
from bitbots_utils.utils import get_parameters_from_other_node
from geometry_msgs.msg import Point
from nav_msgs.msg import OccupancyGrid
from rclpy.node import Node
from ros2_numpy import msgify, numpify
from tf2_geometry_msgs import PointStamped, PoseWithCovarianceStamped


class Map:
    """
    Costmap that keeps track of obstacles like the ball or other robots.
    """

    def __init__(self, node: Node, buffer: tf2.Buffer) -> None:
        self.node = node
        self.buffer = buffer
        self.resolution: int = self.node.config.map.resolution
        parameters = get_parameters_from_other_node(
            self.node, "/parameter_blackboard", ["field.size.x", "field.size.y", "field.size.padding"]
        )
        self.size: tuple[float, float] = (
            parameters["field.size.x"] + 2 * parameters["field.size.padding"],
            parameters["field.size.y"] + 2 * parameters["field.size.padding"],
        )
        self.map: np.ndarray = np.ones(
            (np.array(self.size) * self.resolution).astype(int),
            dtype=np.int8,
        )

        self.frame: str = self.node.config.map.planning_frame
        self.ball_buffer: list[Point] = []
        self.robot_buffer: list[sv3dm.Robot] = []
        self.config_ball_diameter: float = self.node.config.map.ball_diameter
        self.config_inflation_blur: int = self.node.config.map.inflation.blur
        self.config_inflation_dialation: int = self.node.config.map.inflation.dialate
        self.config_obstacle_value: int = self.node.config.map.obstacle_value
        self.ball_obstacle_active: bool = True

    def set_ball(self, ball: PoseWithCovarianceStamped) -> None:
        """
        Adds a given ball to the ball buffer
        """
        point = PointStamped()
        point.header.frame_id = ball.header.frame_id
        point.point = ball.pose.pose.position
        try:
            self.ball_buffer = [self.buffer.transform(point, self.frame).point]
        except (tf2.ConnectivityException, tf2.LookupException, tf2.ExtrapolationException) as e:
            self.node.get_logger().warn(str(e))

    def _render_balls(self) -> None:
        """
        Draws the ball buffer onto the costmap
        """
        ball: sv3dm.Ball
        for ball in self.ball_buffer:
            cv2.circle(
                self.map,
                self.to_map_space(ball.x, ball.y)[::-1],
                round(self.config_ball_diameter / 2 * self.resolution),
                self.config_obstacle_value,
                -1,
            )

    def set_robots(self, robots: sv3dm.RobotArray):
        """
        Adds a given robot array to the robot buffer
        """
        new_buffer: list[sv3dm.Robot] = []
        robot: sv3dm.Robot
        for robot in robots.robots:
            point = PointStamped()
            point.header.frame_id = robots.header.frame_id
            point.point = robot.bb.center.position
            try:
                robot.bb.center.position = self.buffer.transform(point, self.frame).point
                new_buffer.append(robot)
            except (tf2.ConnectivityException, tf2.LookupException, tf2.ExtrapolationException) as e:
                self.node.get_logger().warn(str(e))
        self.robot_buffer = new_buffer

    def _render_robots(self) -> None:
        """
        Draws the robot buffer onto the costmap
        """
        robot: sv3dm.Robot
        for robot in self.robot_buffer:
            cv2.circle(
                self.map,
                self.to_map_space(robot.bb.center.position.x, robot.bb.center.position.y)[::-1],
                round(max(numpify(robot.bb.size)[:2]) * self.resolution),
                self.config_obstacle_value,
                -1,
            )

    def to_map_space(self, x: float, y: float) -> tuple[int, int]:
        """
        Maps a point (x, y in meters) to corresponding pixel on the costmap
        """
        return (
            max(0, min(round((x - self.get_origin()[0]) * self.resolution), self.map.shape[0] - 1)),
            max(0, min(round((y - self.get_origin()[1]) * self.resolution), self.map.shape[1] - 1)),
        )

    def from_map_space_np(self, points: np.ndarray) -> np.ndarray:
        """
        Maps an array of pixel coordinates from the costmap to world coordinates (meters)
        """
        return points / self.resolution + self.get_origin()

    def get_origin(self) -> np.ndarray:
        """
        Origin of the costmap in meters
        """
        return np.array([-self.map.shape[0] / self.resolution / 2, -self.map.shape[1] / self.resolution / 2])

    def clear(self) -> None:
        """
        Clears the complete cost map
        """
        self.map[...] = 1

    def inflate(self) -> None:
        """
        Applies inflation to all occupied areas of the costmap
        """
        idx = self.map == 1
        map = cv2.dilate(
            self.map.astype(np.uint8),
            np.ones((self.config_inflation_dialation, self.config_inflation_dialation), np.uint8),
            iterations=2,
        )
        self.map[idx] = cv2.blur(map, (self.config_inflation_blur, self.config_inflation_blur)).astype(np.int8)[idx]

    def update(self) -> None:
        """
        Regenerates the costmap based on the ball and robot buffer
        """
        self.clear()
        if self.ball_obstacle_active:
            self._render_balls()
        self._render_robots()
        self.inflate()

    def avoid_ball(self, state: bool) -> None:
        """
        Activates or deactivates the ball obstacle
        """
        self.ball_obstacle_active = state

    def get_map(self) -> np.ndarray:
        """
        Returns the costmap as an numpy array
        """
        return self.map

    def get_frame(self) -> str:
        """
        Returns the frame of reference of the map
        """
        return self.frame

    def to_msg(self) -> OccupancyGrid:
        """
        Returns the costmap as an OccupancyGrid message
        """
        msg: OccupancyGrid = msgify(OccupancyGrid, self.get_map().T)
        msg.header.frame_id = self.frame
        msg.info.width = self.map.shape[0]
        msg.info.height = self.map.shape[1]
        msg.info.resolution = 1 / self.resolution
        msg.info.origin.position.x = self.get_origin()[0]
        msg.info.origin.position.y = self.get_origin()[1]
        return msg
