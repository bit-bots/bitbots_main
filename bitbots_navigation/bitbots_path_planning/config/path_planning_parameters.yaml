bitbots_path_planning:
  base_footprint_frame: 
    type: string
    default_value: base_footprint
    description: 'The frame of the robot base'
    read_only: true
  tf_buffer_duration:
    type: double
    default_value: 5.0
    description: '-'
    validation:
      bounds<>: [0.0, 10.0]
  rate:
    type: double
    default_value: 20.0
    description: 'The rate at which the path planning is executed'
    validation:
      bounds<>: [0.0, 100.0]
  
  map:
    planning_frame:
      type: string
      default_value: map
      description: 'The frame in which the path planning is done'
    resolution:
      type: int
      default_value: 20
      description: 'Pixel per meter'
      read_only: true
    size:
      x:
        type: double
        default_value: 11.0
        description: 'The size of the map in x direction'
        read_only: true
      y:
        type: double
        default_value: 8.0
        description: 'The size of the map in y direction'
        read_only: true
    ball_update_topic:
      type: string
      default_value: ball_position_relative_filtered
      description: 'The topic to subscribe to for ball detections'
      read_only: true
    robot_update_topic:
      type: string
      default_value: robots_relative_filtered
      description: 'The topic to subscribe to for robot detections'
      read_only: true
    ball_diameter:
      type: double
      default_value: 0.13
      description: 'The diameter of the ball'
      read_only: true
    obstacle_value:
      type: int
      default_value: 50
      description: 'The value of the obstacles in the map'
      read_only: true
    inflation:
      dialate:
        type: int
        default_value: 3
        description: 'The dialte value for the inflation'
        read_only: true
      blur:
        type: int
        default_value: 13
        description: 'The blur value for the inflation'
        read_only: true

  controller:
    carrot_distance:
      type: int
      default_value: 20
      description: 'The distance to the carrot that we want to reach on the path'
      validation:
        bounds<>: [0, 100]
    max_rotation_vel:
      type: double
      default_value: 0.4
      description: 'The maximum rotation velocity of the robot in rad/s around the z-axis'
      validation:
        bounds<>: [0.0, 1.0]

    max_vel_x:
      type: double
      default_value: 0.15
      description: 'Maximum velocity we want to reach in different directions (base_footprint coordinate system)'
      validation:
        bounds<>: [0.0, 1.0]
    min_vel_x:
      type: double
      default_value: -0.1
      description: 'Minimum velocity we want to reach in different directions (base_footprint coordinate system)'
      validation:
        bounds<>: [-1.0, 0.0]
    max_vel_y:
      type: double
      default_value: 0.1
      description: 'Maximum velocity we want to reach in different directions (base_footprint coordinate system)'
      validation:
        bounds<>: [0.0, 1.0]
    smoothing_tau:
      type: double
      default_value: 0.5
      description: 'This is the time constant of the exponential smoothing filter. The higher the value, the more smoothing is applied. It denotes the time after which a unit step function input signal reaches 63% (1-1/e) of its original strength. In other words, it denotes the time it takes for a new input to be 63% integrated into the output. It is update rate independent!'
      validation:
        bounds<>: [0.0, 1.0]
    rotation_i_factor:
      type: double
      default_value: 0.0
      description: 'Integral gain to counter constant disturbances (not really tested/used)'
      validation:
        bounds<>: [0.0, 1.0]
    rotation_slow_down_factor:
      type: double
      default_value: 0.3
      description: 'Clamped p gain of the rotation controller'
      validation:
        bounds<>: [0.0, 1.0]
    translation_slow_down_factor:
      type: double
      default_value: 0.5
      description: 'Clamped p gain of the translation controller'
      validation:
        bounds<>: [0.0, 1.0]
    orient_to_goal_distance:
      type: double
      default_value: 1.0
      description: 'Distance at which we switch from orienting towards the path to orienting towards the goal poses orientation (in meters)'
      validation:
        bounds<>: [0.0, 10.0]
      