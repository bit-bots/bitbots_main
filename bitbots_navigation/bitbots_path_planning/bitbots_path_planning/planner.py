import numpy as np
import pyastar2d
import tf2_ros as tf2
from geometry_msgs.msg import PoseStamped
from nav_msgs.msg import Path
from rclpy.duration import Duration
from rclpy.node import Node
from rclpy.time import Time
from std_msgs.msg import Header

from bitbots_path_planning.map import Map


class Planner:
    """
    A simple grid based A* interface
    """

    def __init__(self, node: Node, buffer: tf2.Buffer, map: Map) -> None:
        self.node = node
        self.buffer = buffer
        self.map = map
        self.goal: PoseStamped | None = None
        self.path: Path | None = None
        self.base_footprint_frame: str = self.node.config.base_footprint_frame

    def set_goal(self, pose: PoseStamped) -> None:
        """
        Updates the goal pose
        """
        pose.header.stamp = Time(clock_type=self.node.get_clock().clock_type).to_msg()
        self.goal = pose

    def cancel(self) -> None:
        """
        Removes the current goal
        """
        self.goal = None
        self.path = None

    def active(self) -> bool:
        """
        Determine if we have an active goal
        """
        return self.goal is not None

    def step(self) -> Path:
        """
        Generates a new A* path to the goal pose with respect to the costmap
        """
        goal = self.goal

        # Get current costmap
        navigation_grid = self.map.get_map()

        # Get my pose and position on the map
        my_position = self.buffer.lookup_transform(
            self.map.frame, self.base_footprint_frame, Time(), Duration(seconds=0.2)
        ).transform.translation

        # Transform goal pose to map frame if needed
        if goal.header.frame_id != self.map.frame:
            goal = self.buffer.transform(goal, self.map.frame, timeout=Duration(seconds=0.2))

        # Run A* from our current position to the goal position
        path = pyastar2d.astar_path(
            navigation_grid.astype(np.float32),
            self.map.to_map_space(my_position.x, my_position.y),
            self.map.to_map_space(goal.pose.position.x, goal.pose.position.y),
            allow_diagonal=False,
        )

        # Convert the pixel coordinates to world coordinates
        path = self.map.from_map_space_np(path)

        # Build path message
        def to_pose_msg(element):
            pose = PoseStamped()
            pose.pose.position.x = element[0]
            pose.pose.position.y = element[1]
            return pose

        poses = list(map(to_pose_msg, path))

        poses.append(goal)
        self.path = Path(
            header=Header(frame_id=self.map.get_frame(), stamp=self.node.get_clock().now().to_msg()), poses=poses
        )

        return self.path

    def get_path(self) -> Path | None:
        """
        Returns the most recent path
        """
        return self.path
