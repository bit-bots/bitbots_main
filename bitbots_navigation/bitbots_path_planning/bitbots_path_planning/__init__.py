# Setting up runtime type checking for this package
from beartype.claw import beartype_this_package
from rclpy.node import Node

from bitbots_path_planning.path_planning_parameters import bitbots_path_planning as parameters

beartype_this_package()


class NodeWithConfig(Node):
    def __init__(self, name: str) -> None:
        super().__init__(name)
        self.param_listener = parameters.ParamListener(self)
        self.config = self.param_listener.get_params()
