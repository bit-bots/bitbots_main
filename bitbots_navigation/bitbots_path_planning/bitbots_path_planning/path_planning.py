import rclpy
import soccer_vision_3d_msgs.msg as sv3dm
from bitbots_tf_buffer import Buffer
from geometry_msgs.msg import PointStamped, PoseStamped, Twist
from nav_msgs.msg import OccupancyGrid, Path
from rclpy.callback_groups import MutuallyExclusiveCallbackGroup
from rclpy.duration import Duration
from rclpy.executors import MultiThreadedExecutor
from rclpy.node import Node
from std_msgs.msg import Empty

from bitbots_msgs.msg import PoseWithCertaintyStamped
from bitbots_path_planning.controller import Controller
from bitbots_path_planning.map import Map
from bitbots_path_planning.planner import Planner


class PathPlanning(Node):
    """
    A minimal python path planning.
    """

    def __init__(self) -> None:
        super().__init__("bitbots_path_planning")

        # Declare params
        self.declare_parameter("base_footprint_frame", "base_footprint")
        self.declare_parameter("rate", 20.0)

        # We need to create a tf buffer
        self.tf_buffer = Buffer(self, Duration(seconds=self.declare_parameter("tf_buffer_duration", 5.0).value))

        # Create submodules
        self.map = Map(node=self, buffer=self.tf_buffer)
        self.planner = Planner(node=self, buffer=self.tf_buffer, map=self.map)
        self.controller = Controller(node=self, buffer=self.tf_buffer)

        # Subscriber
        self.create_subscription(
            PoseWithCertaintyStamped,
            self.declare_parameter("map.ball_update_topic", "ball_relative_filtered").value,
            self.map.set_ball,
            5,
            callback_group=MutuallyExclusiveCallbackGroup(),
        )
        self.create_subscription(
            sv3dm.RobotArray,
            self.declare_parameter("map.robot_update_topic", "robots_relative_filtered").value,
            self.map.set_robots,
            5,
            callback_group=MutuallyExclusiveCallbackGroup(),
        )
        self.goal_sub = self.create_subscription(
            PoseStamped, "goal_pose", self.planner.set_goal, 5, callback_group=MutuallyExclusiveCallbackGroup()
        )
        self.create_subscription(
            Empty,
            "pathfinding/cancel",
            lambda _: self.planner.cancel(),
            5,
            callback_group=MutuallyExclusiveCallbackGroup(),
        )

        # Publisher
        self.cmd_vel_pub = self.create_publisher(Twist, "cmd_vel", 1)
        self.costmap_pub = self.create_publisher(OccupancyGrid, "costmap", 1)
        self.path_pub = self.create_publisher(Path, "path", 1)
        self.carrot_pub = self.create_publisher(PointStamped, "carrot", 1)

        # Timer that updates the path and command velocity at a given rate
        self.create_timer(
            1 / self.get_parameter("rate").value,
            self.step,
            clock=self.get_clock(),
            callback_group=MutuallyExclusiveCallbackGroup(),
        )

    def step(self) -> None:
        """
        Performs a single step of the path planning
        """
        try:
            # Update the map with the latest ball and robot positions
            self.map.update()
            # Publish the costmap for visualization
            self.costmap_pub.publish(self.map.to_msg())

            if self.planner.active():
                # Calculate the path to the goal pose considering the current map
                path = self.planner.step()
                # Publish the path for visualization
                self.path_pub.publish(path)
                # Calculate the command velocity to follow the given path
                cmd_vel, carrot_point = self.controller.step(path)
                # Publish the walk command to control the robot
                self.cmd_vel_pub.publish(cmd_vel)
                # Publish the carrot point for visualization
                self.carrot_pub.publish(carrot_point)
        except Exception as e:
            self.get_logger().error(f"Caught exception during calculation of path planning step: {e}")


def main(args=None):
    rclpy.init(args=args)
    node = PathPlanning()

    # choose number of threads by number of callback_groups + 1 for simulation time
    ex = MultiThreadedExecutor(num_threads=7)
    ex.add_node(node)
    ex.spin()

    node.destroy_node()
    rclpy.shutdown()
