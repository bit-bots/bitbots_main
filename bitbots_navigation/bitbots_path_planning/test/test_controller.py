import math
from unittest.mock import Mock

import pytest
import tf2_ros as tf2
from bitbots_path_planning.controller import Controller
from bitbots_path_planning.path_planning_parameters import bitbots_path_planning as gen_params
from geometry_msgs.msg import Twist
from nav_msgs.msg import Path
from rclpy.node import Node
from rclpy.time import Time
from tf2_geometry_msgs import PoseStamped


def test_default_setup(snapshot, node, tf2_buffer, config):
    controller = setup_controller(node, tf2_buffer)
    parameter_keys = [
        "carrot_distance",
        "max_rotation_vel",
        "max_vel_x",
        "max_vel_y",
        "min_vel_x",
        "orient_to_goal_distance",
        "rotation_slow_down_factor",
        "rotation_i_factor",
        "smoothing_tau",
        "translation_slow_down_factor",
    ]
    parameters = {p: getattr(config.controller, p) for p in parameter_keys}

    assert controller.angular_error_accumulator == 0
    assert controller.last_cmd_vel == Twist()
    assert parameters == snapshot


def test_step_limits_forward_x_velocity(node, tf2_buffer, config, pose_opponent_goal):
    controller = setup_controller(node, tf2_buffer)

    controller.step(path_to(pose_opponent_goal))

    assert controller.last_cmd_vel.linear.x == pytest.approx(config.controller.max_vel_x)
    assert controller.last_cmd_vel.linear.y == pytest.approx(0)
    assert controller.last_cmd_vel.linear.z == 0


def test_step_limits_backward_x_velocity(node, tf2_buffer, config, pose_own_goal):
    controller = setup_controller(node, tf2_buffer)

    controller.step(path_to(pose_own_goal))

    assert controller.last_cmd_vel.linear.x == pytest.approx(config.controller.min_vel_x)
    assert controller.last_cmd_vel.linear.y == pytest.approx(0)
    assert controller.last_cmd_vel.linear.z == 0


def test_step_limits_forward_y_velocity(node, tf2_buffer, config, pose_left_line):
    controller = setup_controller(node, tf2_buffer)

    controller.step(path_to(pose_left_line))

    assert controller.last_cmd_vel.linear.x == pytest.approx(0)
    assert controller.last_cmd_vel.linear.y == pytest.approx(config.controller.max_vel_y)
    assert controller.last_cmd_vel.linear.z == 0


def test_step_limits_backward_y_velocity(node, tf2_buffer, config, pose_right_line):
    controller = setup_controller(node, tf2_buffer)

    controller.step(path_to(pose_right_line))

    assert controller.last_cmd_vel.linear.x == pytest.approx(0)
    assert controller.last_cmd_vel.linear.y == pytest.approx(-config.controller.max_vel_y)
    assert controller.last_cmd_vel.linear.z == 0


def test_step_limits_forward_xy_velocities(node, tf2_buffer, config, pose_opponent_corner):
    controller = setup_controller(node, tf2_buffer)

    controller.step(path_to(pose_opponent_corner))

    goal_heading_angle = math.atan2(pose_opponent_corner.pose.position.y, pose_opponent_corner.pose.position.x)
    walk_command_angle = math.atan2(controller.last_cmd_vel.linear.y, controller.last_cmd_vel.linear.x)

    assert goal_heading_angle == pytest.approx(walk_command_angle)

    walk_command_speed = math.hypot(controller.last_cmd_vel.linear.x, controller.last_cmd_vel.linear.y)

    assert walk_command_speed < config.controller.max_vel_x
    assert walk_command_speed < config.controller.max_vel_y
    assert controller.last_cmd_vel.linear.z == 0


def test_step_limits_backward_xy_velocities(node, tf2_buffer, config, pose_own_corner):
    controller = setup_controller(node, tf2_buffer)

    controller.step(path_to(pose_own_corner))

    goal_heading_angle = math.atan2(pose_own_corner.pose.position.y, pose_own_corner.pose.position.x)
    walk_command_angle = math.atan2(controller.last_cmd_vel.linear.y, controller.last_cmd_vel.linear.x)

    assert goal_heading_angle == pytest.approx(walk_command_angle)

    walk_command_speed = math.hypot(controller.last_cmd_vel.linear.x, controller.last_cmd_vel.linear.y)

    assert walk_command_speed < abs(config.controller.min_vel_x)
    assert walk_command_speed < config.controller.max_vel_y
    assert controller.last_cmd_vel.linear.z == 0


def test_step_limits_rotation(node, tf2_buffer, config, pose_left_line, pose_right_line):
    controller = setup_controller(node, tf2_buffer)

    controller.step(path_to(pose_left_line))
    assert controller.last_cmd_vel.angular.z == pytest.approx(config.controller.max_rotation_vel)

    controller.last_update_time = None
    controller.step(path_to(pose_right_line))
    assert controller.last_cmd_vel.angular.z == pytest.approx(-config.controller.max_rotation_vel)


def test_step_cmd_vel_smoothing(snapshot, node, tf2_buffer, config, pose_opponent_corner):
    controller = setup_controller(node, tf2_buffer)
    controller.node.config.controller.smoothing_tau = 0.5
    controller.last_update_time = Time(seconds=0)

    controller.last_cmd_vel.linear.x = config.controller.max_vel_x
    controller.last_cmd_vel.linear.y = config.controller.max_vel_y
    controller.last_cmd_vel.angular.z = config.controller.max_rotation_vel

    controller.step(path_to(pose_opponent_corner))

    assert str(controller.last_cmd_vel) == snapshot


def setup_controller(node: Node, buffer: tf2.Buffer) -> Controller:
    return Controller(node, buffer)


def path_to(pose: PoseStamped) -> Path:
    path = Path()
    path.header.frame_id = "map"
    path.poses = [pose]

    return path


@pytest.fixture
def tf2_buffer(pose_center_point: PoseStamped) -> tf2.Buffer:
    buffer = Mock(tf2.Buffer)
    buffer.transform.return_value = pose_center_point

    return buffer


@pytest.fixture
def pose_center_point() -> PoseStamped:
    pose = PoseStamped()
    pose.pose.position.x = 0.0
    pose.pose.position.y = 0.0
    pose.pose.position.z = 0.0

    return pose


@pytest.fixture
def pose_own_goal() -> PoseStamped:
    pose = PoseStamped()
    pose.pose.position.x = -4.5
    pose.pose.position.y = 0.0
    pose.pose.position.z = 0.0

    return pose


@pytest.fixture
def pose_opponent_goal() -> PoseStamped:
    pose = PoseStamped()
    pose.pose.position.x = 4.5
    pose.pose.position.y = 0.0
    pose.pose.position.z = 0.0

    return pose


@pytest.fixture
def pose_own_corner() -> PoseStamped:
    pose = PoseStamped()
    pose.pose.position.x = -4.5
    pose.pose.position.y = -3.0
    pose.pose.position.z = 0.0

    return pose


@pytest.fixture
def pose_opponent_corner() -> PoseStamped:
    pose = PoseStamped()
    pose.pose.position.x = 4.5
    pose.pose.position.y = 3.0
    pose.pose.position.z = 0.0

    return pose


@pytest.fixture
def pose_right_line() -> PoseStamped:
    pose = PoseStamped()
    pose.pose.position.x = 0.0
    pose.pose.position.y = -3.0
    pose.pose.position.z = 0.0

    return pose


@pytest.fixture
def pose_left_line() -> PoseStamped:
    pose = PoseStamped()
    pose.pose.position.x = 0.0
    pose.pose.position.y = 3.0
    pose.pose.position.z = 0.0

    return pose


@pytest.fixture
def config() -> gen_params.Params:
    return gen_params.Params()


@pytest.fixture
def node(config) -> Node:
    node = Mock(Node)

    node.config = config

    node.get_logger.return_value.debug = print
    node.get_clock.return_value.now.return_value = Time(seconds=1)

    return node
