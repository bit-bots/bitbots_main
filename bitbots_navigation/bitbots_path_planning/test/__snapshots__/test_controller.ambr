# serializer version: 1
# name: test_default_setup
  dict({
    'carrot_distance': 4,
    'max_rotation_vel': 0.4,
    'max_vel_x': 0.05,
    'max_vel_y': 0.05,
    'min_vel_x': -0.05,
    'orient_to_goal_distance': 1.0,
    'rotation_i_factor': 0.0,
    'rotation_slow_down_factor': 0.3,
    'smoothing_tau': 0.5,
    'translation_slow_down_factor': 0.5,
  })
# ---
# name: test_step_cmd_vel_smoothing
  'geometry_msgs.msg.Twist(linear=geometry_msgs.msg.Vector3(x=0.03270670566473225, y=0.02406005849709838, z=0.0), angular=geometry_msgs.msg.Vector3(x=0.0, y=0.0, z=0.20666164469042259))'
# ---
